### ISSUE TITLE
fixdfti test fails if float generation gets fixed

### ISSUE THREAD

**Issue by est31** [rust-lang/compiler-builtins#162]
In build.rs, there is code to use the random generator to create floats.

It has a little bug though in [the way it determines the exponent](https://github.com/rust-lang-nursery/compiler-builtins/blob/280d19f1127aa80739f4179152b11a5f7d36d79f/build.rs#L3330-L3331). The exponent is first &ed with the mask, and *then* shifted. The issue is though, the mask is already assuming a shifted value. So we end up with the expression always being 0.

Instead it should be first shifted, and then &ed with the mask, like `((exponent << (SIGNIFICAND_BITS -1)) & EXPONENT_MASK) |`.

When doing this though, I discovered that the `fixdfti` test fails. I first thought it was a conversion my PR added, but its in fact already present.

The error looks like:

```
thread 'fixdfti' panicked at 'assertion failed: `(left == right)` (left: `((5179139571476070400,), -170141183460469231731687303715884105728)`, right: `((5179139571476070400,), 170141183460469231731687303715884105727)`)'
```

**Comment by est31** [comment]
This might be related to https://github.com/rust-lang/rust/issues/10184

**Comment by AaronKutch** [comment]
This has been fixed by #397
